src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx

```diff
@@ -0,0 +1,231 @@
+    onRemoveItem: (item, removeItem) => removeItem(item),
+  };
+
+  triggerChange = items => {
+    this.props.onChange(items, {});
+    this.props.onBlur(items, {});
+  };
+
+  addItem = data => {
+    const items = [...this.props.value, data];
+    this.triggerChange(items);
+  };
+
+  updateItem = (data, index) => {
+    const items = [...this.props.value];
+    items.splice(index, 1, data);
+    this.triggerChange(items);
+  };
+
+  removeItem = index => {
+    const items = [...this.props.value];
+    items.splice(index, 1);
+    this.triggerChange(items);
+  };
+
+  onSelectDropdownItem = item => {
+    this.props.onAddItem(item, this.addItem);
+  };
+
+  onEditItem = (item, index) => {
+    this.props.onEditItem(item, data => this.updateItem(data, index));
+  };
+
+  onRemoveItem = (item, index) => {
+    this.props.onRemoveItem(item, () => this.removeItem(index));
+  };
+
+  renderItem = (item, index) => {
+    const removeIcon = (
+      <ItemIcon>
+        <InlineSvg src="icon-trash" size="12px" />
+      </ItemIcon>
+    );
+
+    const removeConfirm = this.props.removeConfirm ? (
+      <Confirm
+        priority="danger"
+        confirmText={t('Remove')}
+        {...this.props.removeConfirm}
+        onConfirm={() => this.onRemoveItem(item, index)}
+      >
+        {removeIcon}
+      </Confirm>
+    ) : (
+      React.cloneElement(removeIcon, {
+        onClick: () => this.onRemoveItem(item, index),
+      })
+    );
+
+    return (
+      <Item size="small" key={index}>
+        <ItemLabel>
+          {this.props.renderItem(item)}
+          <ItemIcon onClick={() => this.onEditItem(item, index)}>
+            <InlineSvg src="icon-settings" size="12px" />
+          </ItemIcon>
+          {removeConfirm}
+        </ItemLabel>
+      </Item>
+    );
+  };
+
+  renderDropdown = () => {
+    return (
+      <DropdownAutoComplete
+        {...this.props.addDropdown}
+        alignMenu="left"
+        onSelect={this.onSelectDropdownItem}
+      >
+        {({isOpen}) => (
+          <DropdownButton icon="icon-circle-add" isOpen={isOpen} size="small">
+            {this.props.addButtonText}
+          </DropdownButton>
+        )}
+      </DropdownAutoComplete>
+    );
+  };
+
+  render() {
+    return (
+      <ItemList>
+        {this.props.value.map(this.renderItem)}
+        {this.renderDropdown()}
+      </ItemList>
+    );
+  }
+}
+
+export default class RichListField extends React.PureComponent {
+  static propTypes = {
+    ...InputField.propTypes,
+    ...RichListProps,
+  };
+
+  renderRichList = fieldProps => {
+    const richListProps = getDefinedProps(RichListProps, this.props);
+    const {value, ...props} = fieldProps;
+
+    // We must not render this field until `setValue` has been applied by the
+    // model, which is done after the field is mounted for the first time. To
+    // check this, we cannot use Array.isArray because the value passed in by
+    // the model might actually be an ObservableArray.
```

Comment:
🤔 why?

Reply:
The answer to this question would only unsettle you.

Reply:
😬

Reply:
The idiocracy of how fields interact with the model, instead of a simple top-down data flow:
 1. The form component renders with initial data. Fields are not rendered at this point.
 2. The model initializes its fields. Fields are still not rendered, which means field props are not available.
 3. Fields render. After they mount, they notify the model about their props.
 4. The model adjusts to the props, and calls `setValue`. 

This means that the initial render of the fields is with plain wrong data. This happens in ProjectGeneralSettings, for instance. Oh, and undo is broken for all fields that use `getValue` / `setValue` for a similar reason.